server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -172,7 +176,57 @@ boolean isJoinPending() {
         return pendingOutgoingJoins.iterator().hasNext();
     }
 
-    void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
+    private static class FailedJoinAttempt {
+        private final DiscoveryNode destination;
+        private final JoinRequest joinRequest;
+        private final TransportException exception;
+        private final long timestamp;
+
+        FailedJoinAttempt(DiscoveryNode destination, JoinRequest joinRequest, TransportException exception) {
+            this.destination = destination;
+            this.joinRequest = joinRequest;
+            this.exception = exception;
+            this.timestamp = System.nanoTime();
+        }
+
+        void maybeLogNow() {
```

Comment:
If you make `isSuspiciousTransportException` return the level then this becomes a one-liner so it's probably simpler to inline it.

Reply:
Good idea, done as a part of this commit. https://github.com/elastic/elasticsearch/pull/39756/commits/737e7f8fb00b4f69093780e812dc42c891b4c2fd

Reply:
Maybe inline this? At least rename it to avoid using `maybe` since it always logs at some level or other.